### DISCUSSION TITLE
How to use Java to build plugins?

### DISCUSSION THREAD

**Discussion by HuangJueXiang** [apache/apisix#5808] [Q&A]
目前我看apisix，可以自定义插件，我现在想在apisix做一套授权的插件
插件工作就是：去某（公司）统一认证中心授权  ，相当于可支持不同插件对接不同的公司授权中心

想用自定义java插件，不知道这个java插件是不是走的NIO，如果不是NIO，那能坑多少请求，会阻塞流程，有没有其他办法弄

自定义插件在 apisix -dashboard不显示，而且无法通过界面配置

**Comment by Baoyuantop** [comment]
About custom plugins are not displayed in apisix-dashboard, You can refer to https://apisix.apache.org/zh/docs/dashboard/FAQ#4-after-modifying-the-plugin-schema-or-creating-a-custom-plugin-in-apache-apisix-why-cant-i-find-it-on-the-dashboard

**Comment by juzhiyuan** [comment]
Hi, may I know what exactly your question is? 😄 Using Java to build plugins or?

> 不知道这个java插件是不是走的NIO，如果不是NIO，那能坑多少请求，会阻塞流程，有没有其他办法弄

Hi @leslie-tsang, could you please give some ideas?

**Comment by leslie-tsang** [comment]
After asking @tzssangglass (an expert of apisix-java-runner) for advice, thanks for his patience
> 想用自定义java插件，不知道这个java插件是不是走的NIO

NIO for sure.
> 那能坑多少请求

It depends on the complexity of the request, here is a simple benchmark sample:

> in rewrite scenario

the latency of java-plugin-runner is **about 2.75 times** higher than the proxy rewrite in P90
